### PR TITLE
Save metrics about the validator's vote account

### PR DESCRIPTION
**Pull request by enriquefynn**
Introduce metrics about the vote account. We can track the amount of rewards (balance of the vote account) and if the validator voted in the last slots.

### CODE REVIEW

File: prometheus/src/cluster_metrics.rs [hunk 1]
```diff
@@ -1,12 +1,84 @@
...
+    last_vote: Slot,
+}
+
+impl ValidatorVoteInfo {
+    fn new_from_bank(bank: &Arc<Bank>, identity_pubkey: &Pubkey) -> Option<Self> {
+        let vote_accounts = bank.vote_accounts();
+        let vote_state_default = VoteState::default();
+        vote_accounts
+            .iter()
+            .filter_map(|(&vote_pubkey, (_activated_stake, account))| {
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]: Do we really have to do a linear search through all vote accounts? Couldn’t we get ours in O(1) by looking it up by pubkey?

Alternatively, if we are going to iterate _all_ vote accounts anyway, we might expose metrics about multiple vote accounts. That would allow our RPC nodes and Lido validator to export the status of our primary validator, and vice versa, so they can monitor each other, and we can see when one is offline.

Exposing _all_ vote accounts sounds like a DoS vector though (what if somebody creates a few hundred thousand vote accounts — wouldn’t be too expensive in terms of SOL and now all of the O(n) logic involving vote accounts slows becomes 100× slower ...). If we do something like that, we should add a new CLI flag, maybe `--monitor-vote-account` or something, to specify the ones we are interested in. Let’s not do that right now, we can do it in a follow-up PR.

File: prometheus/src/cluster_metrics.rs [hunk 2]
```diff
@@ -1,12 +1,84 @@
...
+                let vote_state = account.vote_state();
+                let vote_state = vote_state.as_ref().unwrap_or(&vote_state_default);
+                if identity_pubkey != &vote_state.node_pubkey {
+                    return None;
+                }
+                let last_vote = if let Some(vote) = vote_state.votes.iter().last() {
+                    vote.slot
+                } else {
+                    0
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]: Instead of returning 0, we could make the metric itself absent if there is no last vote. That is more accurate, but I think in practice it will also be nicer to work with, if you run a `delta` or `rate` query on this metric, if it jumps to 0 for some reason you get enormous deltas. If the metric is absent, then the rate/delta is just not defined and absent itself.

File: prometheus/src/cluster_metrics.rs [hunk 3]
```diff
@@ -1,12 +1,84 @@
...
+                    .with_label("vote", self.vote_address.to_string())
+                    .at(at)],
+            },
+        )?;
+        // We can use this metric to track if the validator is making progress
+        // by voting on the last slots.
+        write_metric(
+            out,
+            &MetricFamily {
+                name: "solana_cluster_last_vote_slot_count",
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]:
```suggestion
                name: "solana_cluster_last_vote_slot",
```
This is not a count, it is really the slot.

File: prometheus/src/cluster_metrics.rs [hunk 4]
```diff
@@ -1,12 +1,84 @@
...
+                help: "The last slot that the validator voted on",
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]:
```suggestion
                help: "The voted-on slot of the validator's last vote that got included in the chain",
```

Also, if I understand correctly, this is the last voted on slot, *from the point of view of the chain* (because we access it through the vote account from the bank), *as seen from the validator* (because inside the validator, we only know our own view of the chain). So the sequence is:

1. We vote on slot _x_.
2. The leader includes that vote in block _y_.
3. We receive block _y_.

Only at step 3 would we observe that we voted, and only if we are synced to the chain and using the right fork. This is an important metric to have, but it is slightly different from the the last vote we _sent_ (which probably we cannot get from just the bank).

Because it is based on the bank, I think we should split this metric into the different commitment levels as well, like the other bank-based metrics.

File: prometheus/src/cluster_metrics.rs [hunk 5]
```diff
@@ -1,12 +1,84 @@
...
+                type_: "gauge",
+                metrics: vec![Metric::new(self.last_vote).at(at)],
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]: We should include the vote account public key, and if possible the associated identity account, to avoid ambiguity.

Also, if we add CLI flags to specify accounts to monitor, then we could expose those vote accounts' last vote here as well, and the label would disambiguate them.

File: prometheus/src/cluster_metrics.rs [hunk 6]
```diff
@@ -1,12 +1,84 @@
...
+            },
+        )?;
+        // Validator rewards go to vote account, we use this to track our own
+        // rewards.
+        write_metric(
+            out,
+            &MetricFamily {
+                name: "solana_cluster_vote_balance_total",
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]:
```suggestion
                name: "solana_cluster_vote_balance_sol",
```
It’s not a total in the sense of an accumulating counter, it’s a gauge. Also, we should include the unit in the name.

File: prometheus/src/cluster_metrics.rs [hunk 7]
```diff
@@ -1,12 +1,84 @@
...
+                Some(ValidatorVoteInfo {
+                    vote_address: vote_pubkey,
+                    balance: vote_balance,
+                    last_vote,
+                })
+            })
+            .next()
+    }
+
+    fn write_prometheus<W: io::Write>(&self, out: &mut W, at: SystemTime) -> io::Result<()> {
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]: Like in the previous PR, we sample everything just in time as Prometheus requests it, so we don’t need this `at`.

**Review by ruuda — COMMENTED** [review]
Looks mostly good! I have a few small comments.

File: prometheus/src/cluster_metrics.rs (hunk 1 above)
Comment (enriquefynn) [reply prometheus/src/cluster_metrics.rs]: Added issue https://github.com/ChorusOne/solstrate/issues/14

File: prometheus/src/cluster_metrics.rs (hunk 5 above)
Comment (enriquefynn) [reply prometheus/src/cluster_metrics.rs]: I'll add it as a label, later one we can include label for each vote account

File: prometheus/src/cluster_metrics.rs [hunk 8]
```diff
@@ -1,12 +1,85 @@
...
+        let vote_state_default = VoteState::default();
+        vote_accounts
+            .iter()
+            .filter_map(|(&vote_pubkey, (_activated_stake, account))| {
+                let vote_state = account.vote_state();
+                let vote_state = vote_state.as_ref().unwrap_or(&vote_state_default);
+                if identity_pubkey != &vote_state.node_pubkey {
+                    return None;
+                }
+                let last_vote = if let Some(vote) = vote_state.votes.iter().last() {
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]: If votes is a slice, we could get the last vote in O(1) instead of iterating the votes. In theory it could be O(1) with the iterator as well, but I’m not sure what the status of specialization in Rust is, I wouldn’t count on it.

File: prometheus/src/cluster_metrics.rs [hunk 9]
```diff
@@ -1,12 +1,85 @@
...
+        )?;
+        // Validator rewards go to vote account, we use this to track our own
+        // rewards.
+        write_metric(
+            out,
+            &MetricFamily {
+                name: "solana_node_vote_balance_sol",
+                help: "The current node's vote account balance",
+                type_: "gauge",
+                metrics: vec![Metric::new_sol(self.balance)],
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]: We should add a label with the vote account public key as well.

File: prometheus/src/cluster_metrics.rs [hunk 10]
```diff
@@ -1,12 +1,85 @@
...
+        // by voting on the last slots.
+        write_metric(
+            out,
+            &MetricFamily {
+                name: "solana_node_last_vote_slot",
+                help:
+                    "The voted-on slot of the validator's last vote that got included in the chain",
+                type_: "gauge",
+                metrics: vec![
+                    Metric::new(self.last_vote).with_label("pubkey", self.vote_address.to_string())
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]:
```suggestion
                    Metric::new(self.last_vote).with_label("vote_account", self.vote_address.to_string())
```

Ideally we would add the associated identity account as well. With the Lido for Solana alerts, in some cases I wanted to build a message based on the identity account, to be able to link to a block explorer, but we didn’t have that information available in the metrics.

**Review by ruuda — APPROVED** [review]
I have one more request :grimacing: Can we in all places where we label something with a vote account, also label it with the associated identity account? We should have it available anyway, and it will make it easier to filter metrics and display them consistently if we have the option of using either.

Aside from that, LGTM :shipit:

**Comment by enriquefynn** [comment]
I ended up doing the Issue https://github.com/ChorusOne/solstrate/issues/14 in this PR and I went down this rabbit hole.

I wanted to modify the RPC server so that it works both with `solana-validator` and `solana-test-validator`. So I added the field `observable_vote_accounts` to the `RpcRequestMiddleware` at `rpc_service.rs` -> to `JsonRpcService` -> to `ValidatorConfig`.

I ended up changing `for_each_commitment` to receive `Option` and return also an `Option`, in case vote accounts weren't there, or any thing that shouldn't produce a metric. But, would be good to have something like:
If all the metrics (finalized, processed, confirmed) are `None`, return `None`, otherwise, return a vec with the non-None metrics.
I'm still not so convinced about the way this is handled and was hoping @ruuda can have some insights to make it prettier.

I'll leave this for now and work on other things, will come back at it at some point, when you can add some review. But don't make it a priority.

**Comment by ruuda** [comment]
> If all the metrics (finalized, processed, confirmed) are None, return None, otherwise, return a vec with the non-None metrics.

It could happen in theory that an account gets created or deleted, so it exists at some commitment level and not at another. So I think it does make sense to return `Option` and just collect those. It’s the honest reflection of what happened on the chain.

File: core/src/validator.rs [hunk 11]
```diff
@@ -165,6 +165,7 @@ pub struct ValidatorConfig {
     pub no_wait_for_vote_to_start_leader: bool,
     pub accounts_shrink_ratio: AccountShrinkThreshold,
     pub wait_to_vote_slot: Option<Slot>,
+    pub observable_vote_acounts: Arc<HashSet<Pubkey>>,
```
Comment (ruuda) [inline core/src/validator.rs]: Bikeshed alert, 
```suggestion
    pub vote_accounts_to_monitor: Arc<HashSet<Pubkey>>,
```
what does it mean these vote accounts are “observable”? To me that sounds like other vote accounts are somehow not observable?

Also about the `Arc`, it does help to have only a single instance if this is referenced in multiple places, but it’s also a slight complication, a pubkey is only 32 bytes, even if you monitor a thousand of them it’s only a few kilobytes. Perhaps it’s worth just copying the hash set if it’s needed in multiple places.

Though with the `Arc`, it makes the hash set immutable, so at least nobody in the future will try to do something like modify it at runtime and expect it to work. If we copy it, you could modify the copy in one place and then be confused about why the other place doesn’t see the new value.

Neither one is clearly better than the other to me, I have no strong opinion on it.

File: prometheus/src/bank_metrics.rs [hunk 12]
```diff
@@ -15,7 +15,8 @@ pub fn write_bank_metrics<W: io::Write>(
             help: "Block Slot",
             type_: "gauge",
             metrics: banks_with_commitments
-                .for_each_commitment(|bank| Metric::new(bank.clock().slot)),
+                .for_each_commitment(|bank| Some(Metric::new(bank.clock().slot)))
+                .unwrap(),
```
Comment (ruuda) [inline prometheus/src/bank_metrics.rs]: Why does `for_each_commitment` now return something that needs to be unwrapped? And why is that safe to do here? Can’t we make it always succeed? If all calls to the closure return `None`, we would just have no metrics. That’s fine. We'd print a `# HELP` and `# TYPE` line for something that is missing then, but then a human can go and look and hopefully figure out why it’s missing.

File: prometheus/src/banks_with_commitments.rs [hunk 13]
```diff
@@ -59,11 +59,14 @@ impl BanksWithCommitments {
...
-        ]
+    pub fn for_each_commitment<F: Fn(&Bank) -> Option<Metric>>(
+        &self,
+        get: F,
+    ) -> Option<Vec<Metric>> {
+        Some(vec![
+            get(&self.finalized_bank)?.with_label("commitment_level", "finalized".to_owned()),
+            get(&self.confirmed_bank)?.with_label("commitment_level", "confirmed".to_owned()),
+            get(&self.processed_bank)?.with_label("commitment_level", "processed".to_owned()),
+        ])
```
Comment (ruuda) [inline prometheus/src/banks_with_commitments.rs]: We can do it like this:

```suggestion
    ) -> Vec<Metric> {
        let mut result = Vec::with_capacity(3);
        result.extend(get(&self.finalized_bank).map(|m| m.with_label("commitment_level", "finalized".to_owned()));
        result.extend(get(&self.confirmed_bank).map(|m| m.with_label("commitment_level", "confirmed".to_owned()));
        result.extend(get(&self.processed_bank).map(|m| m.with_label("commitment_level", "processed".to_owned()));
        result
```

It doesn’t have to be all-or-nothing for the metrics.

Comment (ruuda) [reply prometheus/src/banks_with_commitments.rs]: But thinking about it some more ... instead of returning an `Option<Metric>`, shouldn’t we maybe return a `Vec<Metric>` from the closure? If we are going to monitor multiple vote accounts, it will be the same metric, but with different labels. So inside the closure we would loop over the vote account addresses, and append to the vec.

Hmm, and then the closure has to allocate the vec and put things in there, and then we deconstruct the vec again and concatenate it here ... we could do better by defining a wrapper type:

```rust
struct MetricsVec {
    inner: Vec<Metric>,
    commitment_level: String,
}

impl MetricVec {
    pub fn push(&mut self, metric: Metric) {
        self.inner.push(metric.with_label("commitment_level", self.commitment_level.clone()));
    }
}
```
and then the closure would change to `F: Fn(&Bank, &mut MetricsVec)`, and it could put its metrics directly into the right vec. If there are multiple, it can add multiple. If there are none, it just doesn’t push anything.

File: validator/src/main.rs [hunk 14]
```diff
@@ -370,6 +370,22 @@ fn hardforks_of(matches: &ArgMatches<'_>, name: &str) -> Option<Vec<Slot>> {
     }
 }
 
+fn get_observable_vote_accounts(matches: &ArgMatches<'_>) -> HashSet<Pubkey> {
```
Comment (ruuda) [inline validator/src/main.rs]: (bikeshed, but I think this is a bit clearer)
```suggestion
fn get_vote_accounts_to_monitor(matches: &ArgMatches<'_>) -> HashSet<Pubkey> {
```

File: validator/src/main.rs [hunk 15]
```diff
@@ -370,6 +370,22 @@ fn hardforks_of(matches: &ArgMatches<'_>, name: &str) -> Option<Vec<Slot>> {
...
+    let vote_account = if matches.is_present("vote_account") {
+        vec![pubkey_of(&matches, "vote_account").unwrap()]
```
Comment (ruuda) [inline validator/src/main.rs]: I suppose things fail elsewhere already if you pass something invalid to `--vote-account`? But to make it a bit safer, instead of unwrapping, we could preserve the `Option` (or convert it into one with `.ok()` if it was a `Result`), and then in the `else` arm return `None`. `Option` implements `IntoIterator` so it can be used with `.extend` below.

Or even just move the check down, and insert it conditionally.

**Edit**: Ah, the unwrap is safe because Clap already validates the value ... In that case I prefer to use `.expect()` and put the reason for why it is safe in there. (It also helps to keep superficial auditors at bay who say “unwrap is bad, mkay?”.)

File: validator/src/main.rs [hunk 16]
```diff
@@ -1799,6 +1815,15 @@ pub fn main() {
...
                          then this not a good time for a restart")
         )
+        .arg(
+            Arg::with_name("monitor_vote_account")
+            .long("monitor-vote-account")
+            .takes_value(true)
+            .value_name("MONITOR_VOTE_ACCOUNT")
+            .validator(is_pubkey)
+            .multiple(true)
+            .help("The vote accounts to inspect for prometheus metrics")
```
Comment (ruuda) [inline validator/src/main.rs]:
```suggestion
            .help("Additional vote accounts expose Prometheus metrics about. The validator's own vote account is always included implicitly if there is one.")
```

File: validator/src/main.rs [hunk 17]
```diff
@@ -1799,6 +1815,15 @@ pub fn main() {
             .after_help("Note: If this command exits with a non-zero status \
                          then this not a good time for a restart")
         )
+        .arg(
+            Arg::with_name("monitor_vote_account")
+            .long("monitor-vote-account")
+            .takes_value(true)
+            .value_name("MONITOR_VOTE_ACCOUNT")
```
Comment (ruuda) [inline validator/src/main.rs]:
```suggestion
            .value_name("PUBKEY")
```

File: prometheus/src/lib.rs [hunk 18]
```diff
@@ -5,13 +5,16 @@ mod utils;
 
 use banks_with_commitments::BanksWithCommitments;
 use solana_gossip::cluster_info::ClusterInfo;
-use std::sync::Arc;
+use solana_sdk::pubkey::Pubkey;
+use std::{collections::HashSet, sync::Arc};
 
+#[derive(Clone)]
```
Comment (ruuda) [inline prometheus/src/lib.rs]: Could also do `Copy`, it’s only a `u64`.

File: prometheus/src/cluster_metrics.rs [hunk 19]
```diff
@@ -1,15 +1,36 @@
...
     banks_with_commitments::BanksWithCommitments,
     utils::{write_metric, Metric, MetricFamily},
     Lamports,
 };
-use std::{io, sync::Arc};
+use std::{collections::HashSet, io, sync::Arc};
+
+struct ValidatorVoteInfo {
+    balance: Lamports,
+    last_vote: Slot,
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs:15]: Ah, we should also include the vote credits. We use this in the Lido for Solana metrics to see which validators are performing best vote-wise, if you know the vote credits at some time t0 in the past, and the vote credits now, that’s how many votes the validator got included in the chain.

(Sorry for the scope creep :grimacing:)

File: prometheus/src/cluster_metrics.rs [hunk 20]
```diff
@@ -35,10 +56,14 @@ pub fn write_cluster_metrics<W: io::Write>(
             name: "solana_node_identity_balance_sol",
             help: "The node's finalized identity balance",
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]:
```suggestion
            help: "The balance of the node's identity account",
```

File: prometheus/src/cluster_metrics.rs [hunk 21]
```diff
@@ -52,5 +77,50 @@ pub fn write_cluster_metrics<W: io::Write>(
         },
     )?;
 
+    // Vote accounts information
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs:86]: Hmm yeah, so now we run into this issue that we have the structure of our metrics “inside out” ... It is a bit ugly indeed ... I need to think about it a bit more, I have a feeling that an elegant solution exists.

File: prometheus/src/cluster_metrics.rs [hunk 22]
```diff
@@ -52,5 +81,60 @@ pub fn write_cluster_metrics<W: io::Write>(
...
+    for vote_account in vote_accounts.iter() {
+        write_metric(
+            out,
+            &MetricFamily {
+                name: "solana_node_last_vote_slot",
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]: Now that we are going to monitor multiple account, I think the `solana_node_*` series of metrics might become confusing ... on the one hand the identity of the instance itself is `solana_node_identity_public_key_info`, but we also have “nodes” that are not *this* node ...

Maybe we should call these accounts that we monitor `solana_validator_*`?

File: prometheus/src/cluster_metrics.rs [hunk 23]
```diff
@@ -52,5 +81,60 @@ pub fn write_cluster_metrics<W: io::Write>(
...
+                    )
+                }),
+            },
+        )?;
+
+        write_metric(
+            out,
+            &MetricFamily {
+                name: "solana_node_vote_balance_sol",
+                help: "The current node's vote account balance",
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]:
```suggestion
                help: "The balance of the vote account at the given address",
```
It’s not just the current node any more!

File: prometheus/src/cluster_metrics.rs [hunk 24]
```diff
@@ -52,5 +81,60 @@ pub fn write_cluster_metrics<W: io::Write>(
...
+            &MetricFamily {
+                name: "solana_node_last_vote_slot",
+                help:
+                    "The voted-on slot of the validator's last vote that got included in the chain",
+                type_: "gauge",
+                metrics: banks_with_commitments.for_each_commitment(|bank| {
+                    let vote_info = get_vote_state(bank, vote_account)?;
+                    Some(
+                        Metric::new(vote_info.last_vote)
+                            .with_label("identity_account", identity_pubkey.to_string())
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]: This identity pubkey is defined at the beginning of the scope, it is _our_ identity (of the running process), not the one that goes with the vote account! I think we can get the identity from `VoteState::node_pubkey`.

File: prometheus/src/cluster_metrics.rs [hunk 25]
```diff
@@ -52,5 +81,60 @@ pub fn write_cluster_metrics<W: io::Write>(
...
+                    )
+                }),
+            },
+        )?;
+
+        write_metric(
+            out,
+            &MetricFamily {
+                name: "solana_node_vote_credits",
+                help: "The current node's vote vote credits for current epoch",
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]:
```suggestion
                help: "The total number of vote credits credited to this vote account",
```
Also here, this is no longer specific to this node, and also the credits are not specific to the current epoch, they are the total since the beginning of time. I see that the `VoteState` representation changed to also track it since the beginning of the epoch, but this value is still the vote credits since the beginning of time.

(And that is also more useful to us: once Prometheus has a time series of this, we can compute the number of vote credits gained between any two points in time, not just the start of the epoch.)

File: rpc/src/rpc_service.rs [hunk 26]
```diff
@@ -74,6 +74,7 @@ struct RpcRequestMiddleware {
     bank_forks: Arc<RwLock<BankForks>>,
     health: Arc<RpcHealth>,
     block_commitment_cache: Arc<RwLock<BlockCommitmentCache>>,
+    observable_vote_accounts: Arc<HashSet<Pubkey>>,
```
Comment (ruuda) [inline rpc/src/rpc_service.rs]:
```suggestion
    vote_accounts_to_monitor: Arc<HashSet<Pubkey>>,
```
?

File: prometheus/src/cluster_metrics.rs (hunk 25 above)
Comment (enriquefynn) [reply prometheus/src/cluster_metrics.rs]: So this is cumulative?https://github.com/ChorusOne/solana/blob/50604103de3c90f8a5aac135bfe13b0a7ccd614f/programs/vote/src/vote_state/mod.rs#L1047

The comment on `VoteState` says:
```rust
/// history of how many credits earned by the end of each epoch
/// each tuple is (Epoch, credits, prev_credits)
```
I think I got confused here. Testing it with `solana-test-validator` showed me that this is indeed accumulating 👍

File: prometheus/src/cluster_metrics.rs (hunk 24 above)
Comment (enriquefynn) [reply prometheus/src/cluster_metrics.rs]: Good catch!

File: prometheus/src/cluster_metrics.rs (hunk 25 above)
Comment (ruuda) [reply prometheus/src/cluster_metrics.rs]: Thanks for confirming!

File: prometheus/src/cluster_metrics.rs [hunk 27]
```diff
@@ -1,15 +1,44 @@
...
+    balance: Lamports,
+    last_vote: Slot,
+    vote_credits: u64,
+    identity: Pubkey,
+}
+
+fn get_vote_state(bank: &Bank, vote_pubkey: &Pubkey) -> Option<ValidatorVoteInfo> {
+    let default_vote_state = VoteState::default();
+    let vote_accounts = bank.vote_accounts();
+    let (_activated_stake, vote_account) = vote_accounts.get(vote_pubkey)?;
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs:23]: Ah yes, stake! Why not expose that as well? It’s one of our most important metrics :sweat_smile: 

But let’s do it in a follow-up PR, I’d like to get this one merged.

File: prometheus/src/cluster_metrics.rs [hunk 28]
```diff
@@ -52,5 +83,60 @@ pub fn write_cluster_metrics<W: io::Write>(
...
+                            .with_label("vote_account", vote_account.to_string()),
+                    )
+                }),
+            },
+        )?;
+
+        write_metric(
+            out,
+            &MetricFamily {
+                name: "solana_validator_vote_balance_sol",
```
Comment (ruuda) [inline prometheus/src/cluster_metrics.rs]:
```suggestion
                name: "solana_validator_vote_account_balance_sol",
```

**Review by ruuda — APPROVED** [review]
LGTM with one final nit, let’s get this merged!

The querying-the-bank-multiple-times is probably not an issue in practice, and the code look fine, it’s more of an optimization itch than a maintainability issue, so let’s go with it for now :+1:
